packages/compiler/src/compiler/scope.ts

```diff
@@ -15,65 +15,54 @@ export default class Scope {
    * In this case, there are two scopes: root and if. Both scopes share the same variable `var1`,
    * and modifying it in the if scope should also modify it in the root scope. But `var2` is only
    * defined in the if scope and should not be accessible in the root scope.
+   * 
+   * Local pointers
+   * Every scope has its own local pointers that contains the indexes of the variables in the global stash.
    */
-  private static stash: unknown[] = [] // Stash of every variable value in every scope
-  private static readFromStash(index: number): unknown {
-    return Scope.stash[index]
-  }
-  private static addToStash(value: unknown): number {
-    Scope.stash.push(value)
-    return Scope.stash.length - 1
-  }
+  private stash: unknown[] = [] // Stash of every variable value in the global scope
```

Comment:
globalScope would be a better name imo

Reply:
I'll change the names to `globalStash` and `localPointers`.
\*All\* variable values are in the `globalStash`, but that doesn't mean they're all accessible from any scope. The values accessible to each scope are defined in the `localPointers`, which point to the actual value in the stash.

Reply:
The stash is not the scope. The scope is what variables (from the stash) you have access to. The stash is where all variables live.

It's like in javascript. Two variables can point to the same value:
```ts
const var1 = { foo: 'bar' }
const var2 = var1

if (true) {
  var1.foo = 'qux'
}

console.log(var2.foo) // 'qux'
```

Modifying a variable inside the `if` statement, also modifies the value from outside.
However, inside the `if` statement there is a different scope. One copied from the old scope, but with different variable access.

```ts
const var1 = { foo: 'bar' }

if (true) {
  const var2 = var1
  var1.foo = 'qux'
}

console.log(var2.foo) // ReferenceError: var2 is not defined
```

Reply:
> The scope is what variables (from the stash) you have access to.

isn't that what local pointer are?

Reply:
talked offline. 

this.globalStash is indeed the global scope of variables but we wanna avoid calling globalScope because carlos thinks it can lead devs to thinking the scope has access to the global scope when in fact it only has access to the scope available via local pointers.

Reply:
It would have been cool to call them heap and pointers as a memory management joke though (don't do it)